File: src/main/java/com/github/rogerbarton/intellijsv/SvSyntaxHighlighter.java
package com.github.rogerbarton.intellijsv;

import com.github.rogerbarton.intellijsv.psi.SvTypes;
import com.intellij.lexer.Lexer;
import com.intellij.openapi.editor.DefaultLanguageHighlighterColors;
import com.intellij.openapi.editor.HighlighterColors;
import com.intellij.openapi.editor.colors.TextAttributesKey;
import com.intellij.openapi.fileTypes.SyntaxHighlighterBase;
import com.intellij.psi.TokenType;
import com.intellij.psi.tree.IElementType;
import org.jetbrains.annotations.NotNull;

import static com.intellij.openapi.editor.colors.TextAttributesKey.createTextAttributesKey;

public class SvSyntaxHighlighter extends SyntaxHighlighterBase {

    public static final TextAttributesKey SEPARATOR =
            createTextAttributesKey("SV_SEPARATOR", DefaultLanguageHighlighterColors.OPERATION_SIGN);
    public static final TextAttributesKey KEY =
            createTextAttributesKey("SV_KEY", DefaultLanguageHighlighterColors.KEYWORD);
    public static final TextAttributesKey VALUE =
            createTextAttributesKey("SV_VALUE", DefaultLanguageHighlighterColors.STRING);
    public static final TextAttributesKey COMMENT =
            createTextAttributesKey("SV_COMMENT", DefaultLanguageHighlighterColors.LINE_COMMENT);
    public static final TextAttributesKey BAD_CHARACTER =
            createTextAttributesKey("SV_BAD_CHARACTER", HighlighterColors.BAD_CHARACTER);

    private static final TextAttributesKey[] BAD_CHAR_KEYS = new TextAttributesKey[]{BAD_CHARACTER};
    private static final TextAttributesKey[] SEPARATOR_KEYS = new TextAttributesKey[]{SEPARATOR};
    private static final TextAttributesKey[] KEY_KEYS = new TextAttributesKey[]{KEY};
    private static final TextAttributesKey[] VALUE_KEYS = new TextAttributesKey[]{VALUE};
    private static final TextAttributesKey[] COMMENT_KEYS = new TextAttributesKey[]{COMMENT};
    private static final TextAttributesKey[] EMPTY_KEYS = new TextAttributesKey[0];

    @Override
    public @NotNull Lexer getHighlightingLexer() {
        return new SvLexerAdapter();
    }

    @Override
    public TextAttributesKey @NotNull [] getTokenHighlights(IElementType tokenType) {
        if (tokenType.equals(SvTypes.SEPARATOR))
            return SEPARATOR_KEYS;
        else if (tokenType.equals(SvTypes.KEY))
            return KEY_KEYS;
        else if (tokenType.equals(SvTypes.VALUE))
            return VALUE_KEYS;
        else if (tokenType.equals(SvTypes.COMMENT))
            return COMMENT_KEYS;
        else if (tokenType.equals(TokenType.BAD_CHARACTER))
            return BAD_CHAR_KEYS;
        else
            return EMPTY_KEYS;
    }
}

File: src/main/kotlin/com/github/rogerbarton/intellijsv/core/SvSyntaxHighlighter.kt
package com.github.rogerbarton.intellijsv.core

import com.github.rogerbarton.intellijsv.core.lexer.SvLexer
import com.github.rogerbarton.intellijsv.core.psi.*
import com.github.rogerbarton.intellijsv.ide.colors.SvColor
import com.github.rogerbarton.intellijsv.psi.SvTypes.*
import com.intellij.lexer.Lexer
import com.intellij.openapi.editor.colors.TextAttributesKey
import com.intellij.openapi.fileTypes.SyntaxHighlighterBase
import com.intellij.psi.TokenType
import com.intellij.psi.tree.IElementType

class SvSyntaxHighlighter : SyntaxHighlighterBase()
{
    override fun getHighlightingLexer(): Lexer = SvLexer()

    override fun getTokenHighlights(tokenType: IElementType): Array<TextAttributesKey> =
        pack(map(tokenType)?.textAttributesKey)

    companion object {
        fun map(tokenType: IElementType): SvColor? = when (tokenType)
        {
            IDENTIFIER -> SvColor.IDENTIFIER

            in SV_KEYWORDS -> SvColor.KEYWORD
            in SV_BUILTIN_TYPES -> SvColor.BUILTIN_TYPE
            SIGNED, UNSIGNED -> SvColor.SIGNING
            in SV_PORT_DIRECTION -> SvColor.PORT_DIRECTION

            in SV_NUMBERS -> SvColor.NUMBER
            LBRACE, RBRACE -> SvColor.BRACES
            LBRACK, RBRACK -> SvColor.BRACKETS
            LPAREN, RPAREN -> SvColor.PARENTHESES

            SEMICOLON -> SvColor.SEMICOLON
            DOT -> SvColor.DOT
            COMMA -> SvColor.COMMA
            in SV_OPERATORS -> SvColor.OPERATOR

            LINE_COMMENT -> SvColor.LINE_COMMENT
            BLOCK_COMMENT -> SvColor.BLOCK_COMMENT

            TokenType.BAD_CHARACTER -> SvColor.BAD_CHARACTER

            else -> null
        }
    }
}